File: backend/apis/complex/read_posts/read_posts.py
from flask import Flask, request, jsonify
from typing import List
from flask_cors import CORS
from datetime import datetime
import os
import sys

import os
import sys

import logging
from logging.handlers import RotatingFileHandler

from requests import get
from invokes import invoke_http


app = Flask(__name__)
CORS(app)

# Configure logging
log_formatter = logging.Formatter('%(asctime)s [%(levelname)s] - %(message)s')
log_handler = RotatingFileHandler(
    'app.log', maxBytes=1024 * 1024, backupCount=5)
log_handler.setFormatter(log_formatter)
app.logger.addHandler(log_handler)
app.logger.setLevel(logging.INFO)

API_URL = f"{os.environ['API_URL']}"
FOLLOW_URL = f"{API_URL}/follow"
POST_URL = f"{API_URL}/post/user_get"
LIKE_URL = f"{API_URL}/like"
USER_URL = f"{API_URL}/user"


@app.route("/read_posts", methods=['GET'])
def read_posts():
    # app.logger.info("THIS IS CALLED")
    try:
        # Send a GET request to retrieve follows information
        print('\n-----Invoking follows microservice-----')

        user_uid = request.headers.get('uid')
        # app.logger.info('test test uid')

        if not user_uid:
            return jsonify({
                "code": 400,
                "message": "User ID not provided in the request headers"
            }), 400

        print('printing user uid')
        print(user_uid)
        new_followsURL = FOLLOW_URL + '/' + user_uid
        follows_response = invoke_http(new_followsURL, method='GET')
        app.logger.info(FOLLOW_URL + '/' + user_uid)

        print('printing follows response')
        print(follows_response)
        print(type(follows_response))
        follows_response["data"] += [user_uid]
        # follows_data = follows_response.get("data", [])
        # print('follows_response:', follows_data)

        # app.logger.info(follows_response)

        # Check if the response is successful
        if "code" not in follows_response or follows_response["code"] != 200:
            # Error handling if the follows microservice call fails
            # app.logger.info('test test')

            return jsonify({
                "code": follows_response.get("code", 500),
                "message": "Failed to retrieve follows information"
            }), follows_response.get("code", 500)

        # Construct a dictionary to store followers for each user
        followers_dict = {}
        followers_dict[user_uid] = follows_response.get("data", [])

        print('printing followers dict')
        print(followers_dict)

        # Retrieve posts for followers of each user

        follower_posts = []

        for user_id, follower_ids in followers_dict.items():
            print(follower_ids)
            for follower_id in follower_ids:
                print(follower_id)
                new_PostsURL = POST_URL + "/" + follower_id
                follower_posts_response = invoke_http(
                    new_PostsURL, method='GET')
                # app.logger.info(follower_posts_response)
                if "code" not in follower_posts_response or follower_posts_response["code"] != 200:
                    # Error handling if the follows microservice call fails
                    continue
                follower_posts += follower_posts_response.get("data", [])

        if not follower_posts:
            return jsonify({
                "code": follower_posts_response.get("code", 500),
                "message": "Failed to retrieve posts information"
            }), follower_posts_response.get("code", 500)

        # Sort user_follower_posts by date
        follower_posts.sort(key=lambda x: x.get(
            "date posted", ""), reverse=True)

        print(follower_posts)
        for idx, follower_post in enumerate(follower_posts):
            post_id = follower_post["post id"]
            poster_id = follower_post["poster id"]
            user_detail = get_user_detail(poster_id)
            likes = get_likes(post_id)
            follower_posts[idx]["likes"] = likes
            follower_posts[idx]["user detail"] = user_detail

        # Return posts for followers of each user
        return jsonify({
            "code": 200,
            "data": follower_posts
        }), 200

    except Exception as e:
        return jsonify({
            "code": 500,
            "message": "Internal server error: " + str(e)
        }), 500


def get_likes(post_id: str) -> List[str]:
    """gets likes for post"""
    post_likes_route = F"{LIKE_URL}/{post_id}"
    response = get(post_likes_route, timeout=5000)
    print(response)
    return response.json()["data"]


def get_user_detail(user_id: str) -> List[str]:
    """gets likes for post"""
    user_detail_route = F"{USER_URL}/{user_id}"
    response = get(user_detail_route, timeout=5000)
    print(response)
    return response.json()["data"]


# Execute this program if it is run as a main script (not by 'import')
if __name__ == "__main__":
    print("This is flask " + os.path.basename(__file__) + " for reading posts...")
    app.run(host="0.0.0.0", port=5120, debug=True)

File: backend/apis/complex/read_posts/read_posts.js
const express = require("express");
const { invokeHttp } = require("./invokes");

const app = express();
const port = 5120;

app.use(express.json());

const API_URL = process.env.API_URL;
const FOLLOW_URL = `${API_URL}/follow`;
const POST_URL = `${API_URL}/post/user_get`;
const USER_URL = `${API_URL}/user`;
const LIKE_URL = `${API_URL}/like`;

app.get("/read_posts", async (req, res) => {
  try {
    const user_uid = req.headers.uid;

    if (!user_uid) {
      return res.status(400).json({
        code: 400,
        message: "User ID not provided in the request headers",
      });
    }

    const new_followsURL = `${FOLLOW_URL}/${user_uid}`;
    const follows_response = await invokeHttp(new_followsURL, "GET");

    if (!follows_response || follows_response.code !== 200) {
      return res.status(500).json({
        code: 500,
        message: "Failed to retrieve follows information",
      });
    }

    const followers_dict = {};
    followers_dict[user_uid] = [...follows_response.data, user_uid] || [];

    let follower_posts = [];
    for (const user_id in followers_dict) {
      const follower_ids = followers_dict[user_id];

      for (const follower_id of follower_ids) {
        const new_PostsURL = `${POST_URL}/${follower_id}`;
        const follower_posts_response = await invokeHttp(new_PostsURL, "GET");
        if (follower_posts_response.code === 200) {
          follower_posts = follower_posts.concat(
            follower_posts_response.data || []
          );
        }
      }
    }
    for (let i = 0; i < follower_posts.length; i++) {
      let followerPost = follower_posts[i];
      let postId = followerPost["post id"];
      let posterId = followerPost["poster id"];
      // Assume getUserDetail and getLikes functions are defined elsewhere
      let userDetail = await invokeHttp(`${USER_URL}/${posterId}`, "GET");
      console.log("user detail:", userDetail);
      let likes = await invokeHttp(`${LIKE_URL}/${postId}`, "GET");
      console.log("likes", likes);
      let allLikes = likes["data"];
      if (allLikes) {
        follower_posts[i]["likes"] = allLikes;
      } else {
        follower_posts[i]["likes"] = [];
      }
      follower_posts[i]["user detail"] = userDetail["data"];
    }

    if (follower_posts.length === 0) {
      return res.status(200).json({
        code: 200,
        data: [],
      });
    }

    follower_posts.sort(
      (a, b) => new Date(b["date posted"]) - new Date(a["date posted"])
    );

    return res.status(200).json({
      code: 200,
      data: follower_posts,
    });
  } catch (error) {
    return res.status(500).json({
      code: 500,
      message: "Internal server error: " + error.message,
    });
  }
});

app.listen(port, () => {
  console.log(`Server is running on port ${port}`);
});
